### ISSUE TITLE
error starting with 5.0.6

### ISSUE THREAD

**Issue by JontAlle** [asamuzaK/withExEditor#110]
I have sent several emails, but they have been blocked, it seems. So now I'm trying over again, here:

After loading 5.0.6 withExEditor (WEE) stopped working, with an error something like:
child process error: Input and output streams "blocked" or not connected.
I dont have the exact error, because it flashes on the screen for a sec or so, and then is gone.

So I ran the Node.js debugger, but the error file shows no errors. Here is the output:
 node /usr/local/src/withExEditorHost-5.0.2/index.js > index.log 2>&1 &
jba@x1ear:/tmp$ cat index.log 
(node:5498) ExperimentalWarning: The fs.promises API is experimental
C{"withexeditorhost":{"message":"getEditorConfig","status":"ready"}}
[1]+  Stopped                 node /usr/local/src/withExEditorHost-5.0.2/index.js > index.log 2>&1

I reinstalled from scratch and rebooted my computer, and went back to the last working version 5.0.2, but nothing works. Same error.
Thunderbird 66
Linux fedora version 29

node -v v10.15.0

**Comment by JontAlle** [comment]
internal/modules/cjs/loader.js:583
    throw err;
    ^

Error: Cannot find module '/home/jba/.config/withexeditorhost/config/firefox/index.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:581:15)
    at Function.Module._load (internal/modules/cjs/loader.js:507:25)
    at Function.Module.runMain (internal/modules/cjs/loader.js:742:12)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:743:3)

I think this explains the problem. The index.js file is not resolved. It is NOT in ~/.config/...

**Comment by JontAlle** [comment]
internal/modules/cjs/loader.js:583
    throw err;
    ^

Error: Cannot find module '/home/jba/.config/withexeditorhost/config/firefox/index.js'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:581:15)
    at Function.Module._load (internal/modules/cjs/loader.js:507:25)
    at Function.Module.runMain (internal/modules/cjs/loader.js:742:12)
    at startup (internal/bootstrap/node.js:283:19)
    at bootstrapNodeJSCore (internal/bootstrap/node.js:743:3)

I think this explains the problem. The index.js file is not resolved. It is NOT in ~/.config/...

**Comment by JontAlle** [comment]
after 5 hrs of work I'm giving up. Looks like there is a path issue. Can find various .js files
Not sure what is going on, and were things are supposed to be located, but the regular install fails,
and it seems the npm run setup runs, but log files say it cant find modules.
It was all working fine until I did an upgrade to 5.0.6. 
Now I cannot use withExEditor: fails with cryptic message.
HELP

**Comment by asamuzaK** [comment]
Are you trying to use withExEditor with Thunderbird?
Thunderbird is not supported yet.

Anyway, you should refresh your host configuration.
Since you already have Node.js v10 installed, it is recommended to get the host from npm.

* Remove "/usr/local/src/withExEditorHost-5.0.2/".
* Remove "/home/jba/.config/withexeditorhost/".
* Install host from npm: `npm i -g withexeditorhost`.
* Move to the host installed path. You can get the global installation path of npm by `npm root -g`. Suppose the path you get is "/usr/local/lib/node_modules", then `cd /usr/local/lib/node_modules/withexeditorhost`.
* Run setup: `node index setup`. Note that the command is NOT `npm run setup`.

For details, see [Host setup from npm](https://github.com/asamuzaK/withExEditorHost#host-setup-from-npm).

**Comment by JontAlle** [comment]
Today you shipped Host-5.0.7-a.1. So I tried again.
1) I removed all traces of the old versions and downloaded the zip file linux-x86_64.zip from 5.0.8-a1
md5sum is:  3f4ec52bf163a83f8c4680b4cf798885  linux-x86_64.zip
index  LICENSE  README.ja.md  README.md
I did chmod +x index, but that will not modify the index file in any way, so I dont think that makes any 
difference.
Then rm -rf ~/.config/withexeditorhost/
This gave an error:npm WARN checkPermissions Missing write access to /usr/lib/node_modules
npm ERR! path /usr/lib/node_modules
..... many more same error
THere is a log file at /home/jba/.npm/_logs/2019-04-01T00_46_53_159Z-debug.log (150 lines long)
SO: then I did:
sudo npm i -g withexeditorhost
and that seemed to work (no errors)
Ah, here are all my debug files. Didn't know about .npm/_logs
ba@x1ear:~/.npm/_logs$ ls -lt
total 40
-rw-r--r--. 1 jba jba 13138 Mar 31 19:46 2019-04-01T00_46_53_159Z-debug.log
-rw-r--r--. 1 jba jba 13183 Mar 29 10:05 2019-03-29T15_05_30_406Z-debug.log
-rw-r--r--. 1 jba jba   712 Mar 28 00:42 2019-03-28T05_42_24_430Z-debug.log
-rw-rw-r--. 1 jba jba   856 Mar 21 14:39 2019-03-21T19_39_00_146Z-debug.log

Back to my steps: 
jba@x1ear:/usr/lib/node_modules$
jba@x1ear:/usr/lib/node_modules$ ls *
npm:
appveyor.yml  bin           changelogs  CONTRIBUTING.md  html  LICENSE   Makefile  node_modules  README.md
AUTHORS       CHANGELOG.md  configure   doc              lib   make.bat  man       package.json  scripts

withexeditorhost:
index.js  LICENSE  modules  node_modules  package.json  README.ja.md  README.md
This is all new to me (which seems good).
pwd: /usr/lib/node_modules/withexeditorhost
jba@x1ear:/usr/lib/node_modules/withexeditorhost$ node index setup

[1] firefox
[2] chrome
[3] chromium
[4] vivaldi
[0] CANCEL

Select a browser. [1...4 / 0]: 1
(node:22692) ExperimentalWarning: The fs.promises API is experimental
Created: /home/jba/.config/withexeditorhost/config/firefox
Created: /home/jba/.config/withexeditorhost/config/firefox/withexeditorhost.sh
Created: /home/jba/.mozilla/native-messaging-hosts/withexeditorhost.json
Input editor path: /usr/bin/vim
Input command line options: 
Created: /home/jba/.config/withexeditorhost/config/firefox/editorconfig.json

so this seemed to work
I'll try it out and get back to you what happens.


Re your question re thunderbird: I edit email (thunderbird) and web pages (firefox). I have been doing
this for a long time, and they both worked find, until 5.0.6, as I previously reported.

**Comment by JontAlle** [comment]
The editor I use with thunderbird is
https://github.com/exteditor/exteditor
I was confused because they look very similar.
I tried to edit a firefox page, and I get the same error as before. There was no log file showing the
errors in ~./npm

I tried once more, logged in as me not sudo:

jba@x1ear:/usr/lib/node_modules$ pwd
/usr/lib/node_modules


jba@x1ear:/usr/lib/node_modules$ npm run setup
npm ERR! path /usr/lib/package.json
npm ERR! code ENOENT
npm ERR! errno -2
npm ERR! syscall open
npm ERR! enoent ENOENT: no such file or directory, open '/usr/lib/package.json'
npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! enoent 

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/jba/.npm/_logs/2019-04-01T01_41_41_220Z-debug.log

here is the contents of the log file mentioned above:
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'run', 'setup' ]
2 info using npm@6.4.1
3 info using node@v10.15.0
4 verbose stack Error: ENOENT: no such file or directory, open '/usr/lib/package.json'
5 verbose cwd /usr/lib/node_modules
6 verbose Linux 4.20.16-200.fc29.x86_64
7 verbose argv "/usr/bin/node" "/usr/bin/npm" "run" "setup"
8 verbose node v10.15.0
9 verbose npm  v6.4.1
10 error path /usr/lib/package.json
11 error code ENOENT
12 error errno -2
13 error syscall open
14 error enoent ENOENT: no such file or directory, open '/usr/lib/package.json'
15 error enoent This is related to npm not being able to find a file.
16 verbose exit [ -2, true ]

**Comment by asamuzaK** [comment]
> I removed all traces of the old versions and downloaded the zip file linux-x86_64.zip from 5.0.8-a1

You got host from npm, so you don't need binaries anymore.
Remove linux-x86_64.zip and unzipped files.

> I tried to edit a firefox page, and I get the same error as before.

Maybe old process is still running.
Restart your browser.

**Comment by JontAlle** [comment]
I was looking at the web page of the thunderbird editor https://github.com/exteditor/exteditor
and it requires arguments to the editor. Specifically:
Your external editor must run in foreground, i.e. must not return before you close the file.

    NEdit: use "nedit" or "nc -wait"
    gvim: use "gvim --nofork"
    and for vim: use "xterm -e vim"

Is this a possible problem for withExEditor, that I'm not using arguments to vim? I have not tried gvim, however gvim is what I prefer.  I could try it, but I'd rather know for sure, and not waste precious time.

Regarding your comments above. I not only restarted my browser, I rebooted. I have tried all the things you suggest.  After 5 hrs the other day of trying every thing I could think of, I cant keep randomly trying things. I need diagnostics. That is what I sent you above, the error messages.
Its not working, and the problem is connectiong vim to the IO stream. I know linux, but virtually nothing about js. Please tell me how to get useful diag, rather than suggest things I can try. I have tried all the obvious things. 
You say I got the host from npm. This doesnt make sense to me. I got index.js from your website in the form of the zip file. I unzipped and then I had index.js. Run that and it installs the host. Your resp above confuses me.  As I said from over a year ago, your instructions are either hard to follow, or something goes wrong during the install process, or I dont know how ot answer a question, such as the arguments to vim.
It all worked for me for over a year. Now its not working. Why?

**Comment by JontAlle** [comment]
If I "reload" is that not the same as restarting the browser? Reload is quick. Restarting the browser seems like a hack, that upsets my tabs.

**Comment by asamuzaK** [comment]
Sorry for my bad English, but to me, you are messing things yourself.

> ```
> jba@x1ear:/usr/lib/node_modules$ pwd /usr/lib/node_modules
> jba@x1ear:/usr/lib/node_modules$ npm run setup
> ```

Wrong directory and wrong command.
Working directory should be:
```
pwd /usr/lib/node_modules/withexeditorhost
```
And command should be:
```
node index setup
```

That's why the error occured.

> I got index.js from your website in the form of the zip file. I unzipped and then I had index.js.

Those index.js and other unzipped files are the files you don't need anymore.
Remove them.

When you run
```
npm i -g withexeditorhost
```
All files for withExEditorHost are saved under `/usr/lib/node_modules/withexeditorhost`.

Note that you should NOT change any files under `/usr/lib/node_modules/withexeditorhost`.
No chmod, or anything.

> Is this a possible problem for withExEditor, that I'm not using arguments to vim?

Only thing I can say is that it depends on the editor.
In general, withExEditorHost will try to spawn a new process of the editor.
Maybe the new process is spawned, or maybe existing process is re-used.

> If I "reload" is that not the same as restarting the browser?

Reloading withExEditor should be sufficient, restarting the browser is more reliable.

**Comment by JontAlle** [comment]
this looks useful. I'll work on it.  I'm confused by he directions. I think this email proves that I'm confused.
I'll go over the above carefully and get back to you.

**Comment by JontAlle** [comment]
I did exactly what you say above
cd /usr/lib/node_modules/withexeditorhost
node index setup
I get the question about my browser (firefox) the path to vim, etc.
Then I restart the host
and it fails.
I have done exactly the above a large number of times. I can reboot, restart the browser, all of that.
nothing works. I always get that message from vim  "in and output is not a terminal"

**Comment by JontAlle** [comment]
I did exactly what you say above
cd /usr/lib/node_modules/withexeditorhost
node index setup
I get the question about my browser (firefox) the path to vim, etc.
Then I restart the host
and it fails.
I have done exactly the above a large number of times. I can reboot, restart the browser, all of that.
nothing works. I always get that message that vim has no in and output

**Comment by asamuzaK** [comment]
It is a problem of your vim.
Please don't ask me for editor specific issues.
Google `vim read stdin write stdout`, maybe it will help.

**Comment by JontAlle** [comment]
Can you tell me how to get debug information from node and from your host program, so I can debug this problem? 
vim is a pretty standard piece of linux code. Its hard for me to understand why it would be vim that is the problem. Thus debuggin info is critical.

**Comment by asamuzaK** [comment]
You can see error logs from the host in the browser console (Ctrl + Shift + J)
Or open options page and check  "Notify Host Error"

**Comment by asamuzaK** [comment]
FYI, have you tried [textern](https://github.com/jlebon/textern)?
It targets Linux, and in the README it shows how to set vim in various ways.

**Comment by asamuzaK** [comment]
Still in trouble?

**Comment by JontAlle** [comment]
In the last several months I have a new computer (Thinkpad X1) and OS (Fedora), so I thought I would start over from scratch and see if I could get withExEditor (wEE) installed, from scratch.

So I went to
  https://github.com/asamuzaK/withExEditorHost 
and read the instructions. Here is what happened:
Downloaded withexeditor for firefox. Not a problem.
Tried to install "host"
Downloaded https://github.com/asamuzaK/withExEditorHost#host-setup-from-npm
/home/jba/.withexeditor
linux-x86_64.zip
and unzipped.

First I tried https://github.com/asamuzaK/withExEditorHost#host-setup-from-npm
> npm i -g withexeditorhost
Failed with an error.
Next I tried "node index setup"
Both times I got the error:
jba@tbear:~/.withexeditor$ node index setup --browser=firefox
/home/BACKUP/home/jba/.withexeditor/index:1
(function (exports, require, module, __filename, __dirname) { ELF
SyntaxError: Invalid or unexpected token
    at new Script (vm.js:79:7)
    at createScript (vm.js:251:10)
    at Object.runInThisContext (vm.js:303:10)
    at Module._compile (internal/modules/cjs/loader.js:657:28)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:742:12)
    at startup (internal/bootstrap/node.js:283:19)

jba@tbear:~/.withexeditor$ node -v
v10.15.0

Then sort of randomly tried things, based on confusing instructions:
https://github.com/asamuzaK/withExEditorHost#host-setup-from-npm
npm root -g
then
sudo npm i -g withexeditorhost
[sudo] password for jba:
+ withexeditorhost@5.0.12
updated 1 package in 1.688s

This seemed to work but it never asked for the path to the editor!
jba@tbear:~/.withexeditor$ node index setup --browser=firefox
/home/BACKUP/home/jba/.withexeditor/index:1
(function (exports, require, module, __filename, __dirname) { ELF
SyntaxError: Invalid or unexpected token
    at new Script (vm.js:79:7)
    at createScript (vm.js:251:10)
    at Object.runInThisContext (vm.js:303:10)
    at Module._compile (internal/modules/cjs/loader.js:657:28)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Function.Module.runMain (internal/modules/cjs/loader.js:742:12)
    at startup (internal/bootstrap/node.js:283:19)

This seemed to work. Not tested yet however.

It never asked for editor!

Error with permissions in:
/usr/lib/node_modules/withexeditorhost
I fixed the permissions owner and mods (which should NOT be necessary)
Then I ran "npm run setup" once again
npm ERR! enoent ENOENT: no such file or directory, open '/home/BACKUP/home/jba/.npm/_logs/package.json'

I then I copied all the files from
/usr/lib/node_modules/withexeditorhost
to
/home/jba/withexeditorhost

This now looks like:
jba@tbear:~/withexeditor$ ls
index.js  LICENSE  modules  node_modules  package.json  README.ja.md  README.md

Now I again, as jba, I run "npm run setup"
This time it gives me the questions it promised (editor path, browser, etc).
It looks happy. Now I need to try it.

So now I have had a week of trying it, and there sill is a major problem. I can edit a page but it loads
old pages into vi. I must give the vi command "n" which reads in the next page. Eventually I get the page
I wanted to edit, and when I do the edit and save, it works.

If I delete /tmp/withExEditor (the file structure with previous edits), it works the first time, but then degenerates after a few edits of pages, as described above.

So in conclusion:
1) The instructions are either wrong or confusiong.
2) With enough "screwing around" I sort of got it working.
3) I suspect the problem remains what to give gvim/vi as the command line. There is no guidance on this quesiton, so it is stil not fully working since it sees more than one file (old files in the "cache" are read in).
4) So at least I can use it if I have some heavy duty edit i need to make, but I would not call it satisfactory. And the install instructions being in error, is obviously a major problem for new users.

AND:
I just tried using wEE to read in this page, and I got a blank page. I then gave the command "n" which
means next file to vi, and it worked. So I'm doing this edit with vi.

Apr 28, 11:30 AM

**Comment by JontAlle** [comment]
withExEditor was working, and today its not listed in my firefox extentions.
I tried to reinstall from
https://addons.mozilla.org/en-US/firefox/addon/withexeditor/
and I get the message:
"Download Failed"
Is there an alternative way to install the plugin?

**Comment by JontAlle** [comment]
Problem solved.

**Comment by OrionRandD** [comment]
@JontAlle @asamuzaK 
Hi Jont and AsamuzaK,
I am facing the same problem here for about 4 weeks.
I have installed using all possible ways, rebooted, restart firefox, but I also get this weird error:
"vim: warning output is not to a terminal" 
"vim: warning input is not from a terminal"

When I set the host up I choose the correct gvim path and everything as before.

Some time ago it used to work well.
But, for more than 4 weeks I got this weird error.
I am sure that the problem is not the (G)vim editor.
I have used this host and extension for a long time before without any error.

AsamuzaK could you try to reproduce this error in a linux machine, if you are not a 
linux guy, you can setup a linux (my distro is siduction debian) box using virtualbox. So, that you can fix the host code...

JontAlle could you write the fixing steps here for me to try a fix?

Thx a lot
By the way, I am using wasavi extension to write this with a little embedded vim like app. But, I really prefer gvim running with withexeditorhost
